### ISSUE TITLE
Base64 Encode HMAC password

### ISSUE THREAD

**Issue by ChrisMcKee** [BcryptNet/bcrypt.net#17]
From #15 
B64 encode hmac before cryptraw; list of libs doing same in issue
Need to use one as an example to check bcrypt b64 vs .net b64

**Comment by ChrisMcKee** [comment]
In V3

**Comment by snekbaev** [comment]
Hi,

I'm planning to move from https://www.nuget.org/packages/BCrypt-Official/0.1.109 and was hoping for an in-place upgrade. With this change in v3, is there anything special I need to be aware of?

Thank you!

**Comment by ChrisMcKee** [comment]
Hi there the packages are namespace compatible so you can just change packages.
The change in v3 is to the enhanced entropy method (which didn't exist in the old lib) adding base64 encoding

**Comment by snekbaev** [comment]
Hi,
current usage pattern is:
```
var mySalt = BCrypt.Net.BCrypt.GenerateSalt();
return BCrypt.Net.BCrypt.HashPassword(str, mySalt);

and

BCrypt.Net.BCrypt.Verify(str, dbHash);
```
Would you personally suggest switching to enhanced hashing? I get the "the more secure the better", however, from #15 discussion it seems like a good to have thingy, but doesn't really affect much or did I misunderstand something? (given that NULL is not an issue for this library)

Thank you!

**Comment by ChrisMcKee** [comment]
Hey; 

Firstly, you don't need to generate the salt separately its all wrapped up in the method; that signature's basically left around for backwards compatibility (and its also used as part of the tests to create repeatable hashing test vectors).

`BCrypt.HashPassword(str)` 

Is the correct standard usage.

-------------

As for the more entropy option, it's very much up to you; you lose nothing by using it.

The NULL byte question raised, was around password containing null being terminated early and the possibility that it could be caused during the use of passing the hmac bytes directly to the encryptor.

Basically the risk that a null byte would be present in the bytes created during HMAC; and that it could accidentally reduce the keyspace in the final hash. **I added a few tests to prove this wasn't the case**

The reasons behind enhanced entropy, which is the method dropbox use are mentioned here

* https://blogs.dropbox.com/tech/2016/09/how-dropbox-securely-stores-your-passwords/
* https://crypto.stackexchange.com/questions/42415/dropbox-password-security

My addition of base64 encoding the hmac was more a compatibility thing as most other libraries have to perform the extra step.


-------------

tldr: BCrypt is secure on its own; there's varied opinions about the value of adding hmac.

**Comment by ChrisMcKee** [comment]
Released in 3.0.1 and 2.1.4

**Comment by snekbaev** [comment]
Alright, thank you!

I near future I will upgrade to v3 and change one of the calls to be ```BCrypt.HashPassword(str)``` and that should do it then.

Thanks again!

**Comment by ChrisMcKee** [comment]
@snekbaev fml pulled that version

**Comment by snekbaev** [comment]
@ChrisMcKee what happened? I think I'm on 3.1.0

**Comment by ChrisMcKee** [comment]
@snekbaev test run didnt clean down the packages so I published the build with broken origin vector (tool I am did a find and replace... now I remember why I dont do that) 
Tests failed on next run, I'd pushed the package.

**Comment by snekbaev** [comment]
@ChrisMcKee what issues should I expect after upgrading to a newer version if there are any and what possible mitigation workarounds exist?

**Comment by ChrisMcKee** [comment]
@snekbaev 3.1.0 is fine; 3.1.1 is broken (any passwords using 3.1.1 would need resetting) 3.1.2 is fine.
See #24

**Comment by snekbaev** [comment]
@ChrisMcKee phew :) thank you for your hard work!
